File: 3761-minimum-absolute-distance-between-mirror-pairs/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <climits>

using namespace std;

class Solution {
public:
    int minAbsoluteDistance(vector<int>& nums) {
        int n = nums.size();
        int min_dist = INT_MAX;
        // Map stores: reverse(nums[i]) -> most recent index i
        unordered_map<int, int> last_seen_rev;

        for (int j = 0; j < n; ++j) {
            int current_val = nums[j];

            // If the current value matches a reversed value from a previous index i
            if (last_seen_rev.count(current_val)) {
                min_dist = min(min_dist, j - last_seen_rev[current_val]);
            }

            // Calculate reverse of current_val to store for future pairs
            int rev_val = reverseNumber(current_val);
            last_seen_rev[rev_val] = j;
        }

        return (min_dist == INT_MAX) ? -1 : min_dist;
    }

private:
    int reverseNumber(int n) {
        long long rev = 0;
        while (n > 0) {
            rev = rev * 10 + (n % 10);
            n /= 10;
        }
        return (int)rev;
    }
};

File: 3761-minimum-absolute-distance-between-mirror-pairs/solution.py
class Solution:
    def minAbsoluteDistance(self, nums: list[int]) -> int:
        # last_seen_rev maps: reverse(nums[i]) -> index i
        last_seen_rev = {}
        min_dist = float('inf')
        
        for j, val in enumerate(nums):
            # Check if current value matches a previously reversed number
            if val in last_seen_rev:
                min_dist = min(min_dist, j - last_seen_rev[val])
            
            # Calculate the reverse of the current number for future lookups
            # The definition of reverse(120) = 21 is naturally handled by int(str[::-1])
            # or by mathematical digit extraction.
            rev_val = self.reverse_int(val)
            
            # Update the map with the current index. 
            # We overwrite previous indices because we want the minimum distance (closest i).
            last_seen_rev[rev_val] = j
            
        return int(min_dist) if min_dist != float('inf') else -1

    def reverse_int(self, n: int) -> int:
        rev = 0
        while n > 0:
            rev = rev * 10 + (n % 10)
            n //= 10
        return rev